### ISSUE TITLE
[ramda] Why does dissoc uses a generic ?

### ISSUE THREAD

**Issue by JalilArfaoui** [flow-typed/flow-typed#2790]
Actual definition is : 

```js
declare function dissoc<T>(
    key: string,
    ...args: Array<void>
  ): (src: { [k: string]: T }) => { [k: string]: T };
declare function dissoc<T>(
    key: string,
    src: { [k: string]: T }
): { [k: string]: T };
```
I don't get why we want to enforce a unique type `T` a object content ? 

I have a union use case where this causes trouble : 

```js
// @flow
import * as R from 'ramda'

type OtherType = { toto: string }

export type ActionA = {|
  type: string,
  payload: {| somePayload: string |}
|}

export type ActionB = {|
  type: string,
  payload: {| someOtherPayload: OtherType |}
|}

export type Action = ActionA | ActionB

const filterAction = (a: Action): {} => R.dissoc('type')(a)
```

I don't see why Flow errors, and I don't see any problem with this code ..

When I change libdef to the following, I don't have any problem : 

```js
declare function dissoc(
    key: string
  ): (src: { [k: string]: any }) => { [k: string]: any };
  declare function dissoc(
    key: string,
    src: { [k: string]: any }
  ): { [k: string]: any };
```

**Comment by AndrewSouthpaw** [comment]
Word. It's a thing I find frustrating about trying to type Ramda in a dynamic language.

As much as possible, we want to avoid using `any` as we then lose information. That's not really a viable solution. I've tried using `mixed` with mixed results (haha). Using `*` type inference might be worth exploring, I just haven't yet.

Personally, in my project at work we use `any` but I'm not convinced that's appropriate for the libdef at large.

cc @LoganBarnett who works on this libdef a lot as well.

**Comment by LoganBarnett** [comment]
@AndrewSouthpaw I see what you did there!

@JalilArfaoui The reason your definition works is because `any` essentially _disables_ Flow for any variables typed that way. If you run `flow coverage --color <file>` over the file that consumes `dissoc` in your codebase, you'll find any return result from `dissoc` (or perhaps its parts) are not covered (by default this show as red).

It's possible to achieve 100% type coverage, but it's work. My team has had a lot of luck with it, but we're not without periods where we're scratching our heads going "How the hell do we make types for this one?".

`dissoc` may have been authored before the `$Rest` type was available and documented. Judging by the `...args: Array<void>` I'd say that's almost certainly the case. I thought we got all of those!

This might be a more typesafe version (minus auto-currying for now):

``` javascript
declare function dissoc<T, K: $Keys<T>>(key: K, src: T): $Rest<$Exact<T>, {| [K]: $ElementType<T, K>|}>
```

[See the try](https://flow.org/try/#0PQKgBAAgZgNg9gdzCYAoVATApgYxgQwCcswoBXAOxwBcBLOCsDWgZxbhwB4AVAGjADSALjAASAVgCeLHgD5ZACgDWUkQP4tCOEdwCUI0QCUsLap1EBRAB74ac-gG8APmADaAgLoGLMLAFssCmpuSQAHLB5+AVknAF9ZVGBgMG44AHM03zBqAAsSHDg-AKCwACMseCR8CgwwfFK4ADcSajgwFiwWnPxqN08wVgoAcl6MOFoKNOy87MJaHCV+UrJegCIoODhVgZYAOkTk7DwiEnIqOgYmVnYuPkEDCWk5RRVJNQ0tHX0xY1Nza1sZm4skcLg2cG8vmKwTCETu0TiCTAYFQ1FhYAAgmAALxgByoZHgkSmOaTXgEspEEQUMh+cqEcmxdBo8JgABCOLxFNKVLANLpWAZKKZqAKFFMdREWNxDlImxEQ3BQyWvIArGAmWKJaURBzccw2BwFIrNsq6rowEkwAg4IQlCwAIToLW9HCcg03Y1K-j4C1WgCycDIHTATUFYDdrXanTACjSWF6tGoAH4LbkWrD9qKGBL8AAmKXu65Gk1wM2+y3JAAicBMw16Nrt-AQOXmOR2YCwVnCNCwGF2QA) - I wasn't able to completely get it working, and I think it has to do with my usage of `[K]`. If I used `foo` instead of a dynamic key, it works fine. Am I annotating it incorrectly?

**Comment by LoganBarnett** [comment]
This issue might be related: https://github.com/facebook/flow/issues/6145

**Comment by AndrewSouthpaw** [comment]
Yeah if we're going to correct `dissoc` we'd definitely want to allow the output type to be a subtype of the input type.

**Comment by retyui** [comment]
How about use it:
```js
declare function dissoc<C,K: string,T: Object>(K, T):  $Rest<T, C>

const dissocd = dissoc<{|'b': any|},_,_>("b", { a: 1, b: 2 });
```

**Comment by LoganBarnett** [comment]
[try of @retyui's post](https://flow.org/try/#0PQKgBAAgZgNg9gdzAZwC4CcCWBjVYTABQhAJgKbYwCG6ZYUArgHa6ZxNgmbLJzYA8AYQBcYAPIAjAFYVUAGgDSotFiYBzOQBVRkmbgB8ACgVywmgJSiwAEgBKZNP02nrAUQAeVXEP37i2djRObl5sEjAAXmCePn4AbwByCQTRKiYATwBfOQB9XKMAIgkC0ziwKlEARlMJUQAmMEzzAG5iLhiwgDoqMGBgMAQ4dABrZFIQvhJOiV7+iVoqUfLkMEw8ZAALOAYYEiA)

It indeed seems to work.

Do we have a means of expressing that the type params must be provided? If omitted, we're back to false positives. Something would be better than nothing though.

As an aside, is this `_` type documented anywhere? I couldn't find anything on it.

**Comment by retyui** [comment]
@LoganBarnett 
One of minuses flow.js is his documentation( 
All new features you can find in GitHub repo.

**Comment by LoganBarnett** [comment]
@AndrewSouthpaw I respectfully disagree. While a normal type check will pass, there are a number of ways that `any` can leak into something and cause Flow to happily ignore anything you do with it. The unfortunate state of things is Flow can't directly be commanded to disallow `any` from my experience, but there are type coverage tools you can employ such as `flow-coverage-report` where you can set the threshold at 100% and add it to your build pipeline.

To be clear: I'd really like to _not_ build types here with that kind of tooling in mind, but I feel like our alternative is to build the types incorrectly (what we have now), or disable it from type checking entirely (`any` as you said).

While it's not exactly the same thing, we have historical precedent for requiring consumer-side typing for some uncommon scenarios just to help Flow out. One such example is the `filter` function using a `%checks` predicate (see [this](https://github.com/flow-typed/flow-typed/blob/master/definitions/npm/ramda_v0.x.x/flow_v0.82.x-/test_ramda_v0.x.x_list.js#L170) for how we do it).

Thoughts?

**Comment by AndrewSouthpaw** [comment]
My concern is that people will never think to look at the documentation and realize that they need to provide consumer-side typing, thus gaining false confidence in their code. Is there a way to avoid that?

**Comment by LoganBarnett** [comment]
I hear and share your concern. I don't know how to completely avoid it other than having type coverage checking alongside actual type checking. I've found this necessary for other cases too though :|

**Comment by AndrewSouthpaw** [comment]
Does the `flow-coverage-report` cover this case, where not providing consumer-side typing silently causes the type to be obliterated?

**Comment by LoganBarnett** [comment]
Generally I've found things like this to be the case, but I will put together a small repo that demonstrates it so we can be sure.

**Comment by retyui** [comment]
More better type:

[Try example](https://flow.org/try/#0CYUwxgNghgTiAEA3W8C2BPAIgSwM64HswAueAbwCh5r4AeAFQBp4BlUs+AbQGtTcAXGNgB2AcwC6pevAC+zANKkAJCwCuAI37oADiFpL5IdLlosAfBYAU3I6XnNcMEqwCU7Lr3gChYyfGly8AD0QfAAxACMVDSmZta2XoIiog5OpCxurMGhYQBMFDIUFGAEwgLwBOoAVvAAvOTR1FDEEYyN8OrEAOQRuQDMXW008M6CqiAFANxFJWX8FdURdWhYeIRgll1QgwtVLpMh4VEUlpVVEQB0zfDCqqjqIDDwAD6JPqIvHQQEECBQwvsTmdLp0bncHk9Xt5kp91N9fv9AadFhdnLd7o9PtCxLD4X8AdNiqVymdcssMDh8ERNgAPHZnfaHPJFZFVXJXUjoiFI0kXUHY0Q86rs5xwn74-ZAA)

```js
declare var myDissoc: {
    <T, S: { [k: string]: T }, K: $Subtype<$Keys<S>>>(key: K, src: S): { [k: string]: T }, // #1
    <S>(key: string, src: S): S // #2
}

const obj = {
    a:1,
    b:'123',
    c: true
};

const obj1 = myDissoc('a', obj);// #1

(obj1.a: number | string | boolean);
(obj1.b: number | string | boolean);
(obj1.c: number | string | boolean);

const obj2 = myDissoc('x', obj);// #2

(obj2.a: number);
(obj2.b: string);
(obj2.c: boolean);
```

**Comment by LoganBarnett** [comment]
Don't we want `(obj1.a: number | string | boolean)` to fail because `a` is not on `obj1`, and `(obj1.b: number | string | boolean)` to fail because it is a `string`?

**Comment by LoganBarnett** [comment]
@AndrewSouthpaw I present to you a [repo of shame](https://github.com/LoganBarnett/flow-type-coverage-demo) and `any`.

**Comment by retyui** [comment]
@LoganBarnett I understand, but `dissoc` works also now:

[Updated example current and my type for dissoc](https://flow.org/try/#0CYUwxgNghgTiAEA3W8C2BPAIgSwM64HswAueAbwCh5r4AeAFQBp4BlUs+AbQGtTcAXGNgB2AcwC6pevAC+zANKkAJCwCuAI37oADiFpL5IdLlosAfBYAU3I6XnNcMEqwCU7Lr3gChYyfGly8AD0QfAAxACMVDSmZta2XoIiog5OpCxurMGhYQBMFDIUFGAEwgLwBOoAVvAAvOTR1FDEEYyN8OrEAOQRuQDMXW008M6CqiAFANxFJWX8FdURdWhYeIRgll1QgwtVLpMh4VEUlpVVEQB0zfDCqqjqIDDwAD6JPqIvHQQEECBQwvsTmdLp0bncHk9Xt5kp91N9fv9AadFhdnLd7o9PtCxLD4X8AdNiqVymdcssMDh8ERNgAPHZnfaHPJFZFVXJXUjoiFI0kXUHY0Q86rs5xwn74wHDajtaUhOXyhWKpXKlWHeAUQ7YfhdXAjAjadDwABmMAIqBGqhgcGE8y0ul1MvVoEgsAQRtUwjA-GwpXgwDWRAYcXaNnQfCSYiGNEczg4PHD7z8AQomTjngFSdkhNmJOqfWW-qpGy29OqgKBeY5YIxkLeMNeYoRBIrVT6fM54MxUIjHwbeMR01ZbbRndrAtx4oHjqAA)

Perhaps this is somewhat useful as in the example with [`assoc()`](https://github.com/flow-typed/flow-typed/issues/2983)

**Comment by LoganBarnett** [comment]
@retyui I'm not sure what you mean by works in this context. However it's indeed helpful to see updated try links and such. Keep them coming!

Given
``` javascript
type Obj = {
  a: number,
  b: string,
  c: bool,
}
      
const obj: Obj = {
    a:1,
    b:'123',
    c: true
};

const obj1 = myDissoc('a', obj);// #1

(obj1.a: number | string | boolean);
(obj1.b: number | string | boolean);
(obj1.c: number | string | boolean);
```

This is from your [try](https://flow.org/try/#0CYUwxgNghgTiAEA3W8C2BPAIgSwM64HswAueAbwCh5r4AeAFQBp4BlUs+AbQGtTcAXGNgB2AcwC6pevAC+zANKkAJCwCuAI37oADiFpL5IdLlosAfBYAU3I6XnNcMEqwCU7Lr3gChYyfGly8AD0QfAAxACMVDSmZta2XoIiog5OpCxurMGhYQBMFDLRNPAUWrrwAPLqAFbwALzk0VCkwqqo6iAwjNHqfEli3dTO6gQEEN2FxdQUYATCAvAENaRVtQ2Uxc0RgzS9AOQRuQDMeztDpIKqIAUA3BQzcwtL1RH1aFh4hGCWe1CnizUXDcQuEohRLM8IgA6ZrwVrtTrwAA+iR8omR8BGYxAUGEQPBkKhvThbQ6MAx3mSGKxEBxeLuEJq0Oc8LJFP66JRNLp+Ie834AOquTeGBw+CIPwAHv9nkCQXl7oyhTCWqTOvilbkiX00RrnlrhqNabj8VMSmaKCErdabba7faHSCSiDsPw9rh4LNtOh4AAzGAEVCe1QwODCAVlEAeoo0CigSCwBC+1TCMD8bBzeDAT5EBhxGM2dA65JnLxpcgeYu+KSyCiZDg8KsSGsyO58p41I5vbPi76-GWAttKo4qkkI8koylialGnkM54j4msxGTjkz7Em+edqEstUT1FUrmzzcx6hAA). I've added an explicit `Obj` type for some clarity, but it's not actually required.

Here is the result of running the code from [Ramda's repl](https://ramdajs.com/repl/?v=0.26.0#?const%20obj%20%3D%20%7B%0A%20%20%20%20a%3A1%2C%0A%20%20%20%20b%3A%27123%27%2C%0A%20%20%20%20c%3A%20true%0A%7D%3B%0A%0AR.dissoc%28%27a%27%2C%20obj%29%3B):

``` javascript
const obj = {
    a:1,
    b:'123',
    c: true
};

R.dissoc('a', obj); // {"b": "123", "c": true}
```

That tells me that the expression `(obj1.a: number | string | boolean);` should fail because we know `a` is not on `obj1`. This is primarily what we're looking for in an updated definition for `dissoc`, unless we've decided this is not the route we want to go.

**Comment by LoganBarnett** [comment]
@AndrewSouthpaw I completely agree that we should warn others of its intended usage. It's an atypical usage for sure. I don't know how to _actively_ warn anyone, but we can certainly use gratuitous comments in both the implementation and the tests around it. Would that work for you, or did you have something else in mind?

**Comment by AndrewSouthpaw** [comment]
That's about all I can come up with as well. 🙁 

Also, @retyui's more recent implement doesn't require the use of all that extra type parameter providing, as it seems to me, so it may be there's no longer anything to actually worry about?

**Comment by AndrewSouthpaw** [comment]
... also, it would be interested to test if the proposed solution actually solves the OP's problem.

**Comment by LoganBarnett** [comment]
@AndrewSouthpaw it's possible I'm missing something here but I still have some outstanding concerns with the PR. Generally I don't want to impede incremental improvements so if that's what that is, that's fine and I'll get out of the way of it. What I'm seeing right now is that the type that `dissoc` might work for an object map but it isn't going to give back a specific structure minus a field (see [my comment here](
https://github.com/flow-typed/flow-typed/issues/2790#issuecomment-443478397)).

**Comment by AndrewSouthpaw** [comment]
Oops, you're quite right, I had missed that.
